b2/b2http.py

```diff
@@ -0,0 +1,178 @@
+        self.code = code
+        self.message = message
+
+    def __repr__(self):
+        return "WebApiError(%d, '%s', '%s')" % (self.status, self.code, self.message)
+
+    def __eq__(self, other):
+        return repr(self) == repr(other)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+def _print_exception(e, indent=''):
+    """
+    Used for debugging to print out nested exception structures.
+    """
+    print(indent + 'EXCEPTION', repr(e))
+    print(indent + 'CLASS', type(e))
+    for (i, a) in enumerate(e.args):
+        print(indent + 'ARG %d: %s' % (i, repr(a)))
+        if isinstance(a, Exception):
+            _print_exception(a, indent + '        ')
+
+
+def _translate_errors(fcn):
+    """
+    Calls the given function, turning any exception raised into a WebApiError.
+    """
+    try:
+        response = fcn()
+        if response.status_code not in [200, 206]:
+            # Decode the error object returned by the service
+            error = json.loads(response.content.decode('utf-8'))
+            raise WebApiError(int(error['status']), error['code'], error['message'])
+        return response
+
+    except WebApiError:
+        raise  # pass through exceptions from just above
+
+    except requests.ConnectionError as e0:
+        e1 = e0.args[0]
+        if isinstance(e1, requests.packages.urllib3.exceptions.MaxRetryError):
+            msg = e1.args[0]
+            if 'nodename nor servname provided, or not known' in msg:
+                # Unknown host, or DNS failing.  In the context of calling
+                # B2, this means that something is down between here and
+                # Backblaze, so we treat it like 503 Service Unavailable.
+                raise WebApiError(503, 'unknown_host', 'unable to locate host')
+        elif isinstance(e1, requests.packages.urllib3.exceptions.ProtocolError):
+            e2 = e1.args[1]
+            if isinstance(e2, socket.error):
+                if e2.args[1] == 'Broken pipe':
+                    # Broken pipes are usually caused by the service rejecting
+                    # an upload request for cause, so we use a 400 Bad Request
+                    # code.
+                    raise WebApiError(503, 'broken_pipe', 'unable to send entire request')
+        raise WebApiError(503, 'connection_error', 'unable to connect')
+
+    except Exception as e:
+        # Don't expect this to happen.  To get lots of info for
+        # debugging, call print_exception.
+        # print_exception(e)
+        raise WebApiError(500, 'unknown', repr(e))
+
+
+def post(url, headers, data):
+    """
+    Posts to the given URL, translating any errors into a raised WebApiError.
+
+    :param url: The URL to post to.
+    :param headers: Headers to send.
+    :param data: Data to send: should be bytes.
+    :return: A requests object with the response.
+    """
+    return _translate_errors(lambda: requests.post(url, headers=headers, data=data))
```

Comment:
Found a [StackOverflow article](https://stackoverflow.com/questions/3252228/python-why-is-functools-partial-necessary).  I'm willing to follow Guido on this one.

(And yes, I did do a lot of Lisp programming early in my career.)